### ISSUE TITLE
 Unit test review: BFT

### ISSUE THREAD

**Issue by janhack** [LiskHQ/lisk-sdk#7957]
# [bft_processing.spec.ts](https://github.com/LiskHQ/lisk-sdk/blob/development/framework/test/unit/engine/bft/bft_processing.spec.ts)

1. In this case, it should not make a difference, but generally it would be good to avoid `floor` and `ceil` on floating point numbers and instead only use integer arithmetic. Concretely,  `Math.floor((scenario.config.activeValidators * 2) / 3) + 1;` could be replaced with integer division.
2. I don't understand the cases `expect(bftInfo?.prevoteWeight).toBeUndefined();` and `expect(bftInfo?.precommitWeight).toBeUndefined();`. I am not sure why why expect the weight to be undefined. Isn't it the case that the `bftInfo` array does not contain an object at that height in the first place.

**Comment by shuse2** [comment]
> In this case, it should not make a difference, but generally it would be good to avoid floor and ceil on floating point numbers and instead only use integer arithmetic. Concretely, Math.floor((scenario.config.activeValidators * 2) / 3) + 1; could be replaced with integer division.

As far as the research by @AndreasKendziorra , the division is [deterministic](https://tc39.es/ecma262/#sec-ecmascript-language-types-number-type) although there will be an rounding error when using floating number

> I don't understand the cases expect(bftInfo?.prevoteWeight).toBeUndefined(); and expect(bftInfo?.precommitWeight).toBeUndefined();. I am not sure why why expect the weight to be undefined. Isn't it the case that the bftInfo array does not contain an object at that height in the first place.

This is because `blockBFTInfo` will be removed with batch size * 3 while inserting a block. although it exists in the fixture, the below min height, the value should be undefined

> areHeadersContradicting: These tests only cover the easy cases, but not the cases depending on maxHeightPrevoted and maxHeightGenerated. I guess we assume here those tests are covered by the function areDistinctHeadersContradicting.

yes. those should be covered in `areDistinctHeadersContradicting`.

> 'should store BFT parameters with height maxHeightPrevoted + 1 if blockBFTInfo does not exist': The BFT parameters should be stored at genesisHeight + 1 in this case (note that only after processing the genesis block, blockBFTInfo should be empty and then we actually have maxHeightPrevoted == genesisHeight).

I think in the BFT, it is not considering genesis height. the logic is if the bft info is empty, store maxHeightPrevoted + 1 which only happens in genesis block case?

**Comment by janhack** [comment]
> I think in the BFT, it is not considering genesis height. the logic is if the bft info is empty, store maxHeightPrevoted + 1 which only happens in genesis block case?

No, in this case the genesis height is used, see the function `getCurrentHeight` in https://github.com/LiskHQ/lips/blob/main/proposals/lip-0058.md#getcurrentheight

**Comment by shuse2** [comment]
> No, in this case the genesis height is used, see the function getCurrentHeight in https://github.com/LiskHQ/lips/blob/main/proposals/lip-0058.md#getcurrentheight

I think we missed to update this, ill try to address it in the PR

**Comment by shuse2** [comment]
I was double checking and currently we store genesis height as maxHeightPrevoted in https://github.com/LiskHQ/lisk-sdk/blob/development/framework/src/engine/bft/module.ts#L49-L60

Therefore, when we use in the empty case https://github.com/LiskHQ/lisk-sdk/blob/development/framework/src/engine/bft/method.ts#L222, the maxHeightPrevoted is genesis height
